File: packages/signal-radar-core/src/signal_radar_core/pipeline.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from .analysis_input import build_analysis_input as build_analysis_input_core
from .memory_application import apply_memory_update


@dataclass(frozen=True)
class PipelineResult:
    exit_code: int
    stdout: str
    stderr: str


class PipelineError(RuntimeError):
    def __init__(self, message: str, result: PipelineResult):
        super().__init__(message)
        self.result = result


def pipeline_error(action: str, exc: Exception) -> PipelineError:
    result = PipelineResult(
        exit_code=1,
        stdout="",
        stderr=f"{type(exc).__name__}: {exc}\n",
    )
    return PipelineError(f"{action} failed: {exc}", result)


def build_analysis_input(
    *,
    config_path: str,
    collector_batch_path: str | Path | None = None,
    skill_dir: str | Path | None = None,
) -> PipelineResult:
    del skill_dir
    try:
        result = build_analysis_input_core(
            config_path=config_path,
            collector_batch_path=collector_batch_path,
        )
        return PipelineResult(exit_code=0, stdout=result.to_stdout(), stderr="")
    except Exception as exc:
        raise pipeline_error("build-analysis-input", exc) from exc


def apply_memory(
    *,
    config_path: str,
    summary_path: str | Path,
    skill_dir: str | Path | None = None,
) -> PipelineResult:
    del skill_dir
    try:
        result = apply_memory_update(
            config_path=config_path,
            summary_path=summary_path,
        )
        return PipelineResult(exit_code=0, stdout=result.to_stdout(), stderr="")
    except Exception as exc:
        raise pipeline_error("apply-memory", exc) from exc

File: packages/signal-radar-core/src/pipeline.ts
import { buildAnalysisInput } from "./analysis-input";
import { applyMemoryUpdate } from "./memory-application";

export type PipelineResult = {
  exit_code: number;
  stdout: string;
  stderr: string;
};

export class PipelineError extends Error {
  result: PipelineResult;

  constructor(message: string, result: PipelineResult) {
    super(message);
    this.name = "PipelineError";
    this.result = result;
  }
}

export async function buildAnalysisInputPipeline(options: {
  configPath: string;
  collectorBatchPath?: string | null;
}): Promise<PipelineResult> {
  try {
    const result = await buildAnalysisInput({
      configPath: options.configPath,
      collectorBatchPath: options.collectorBatchPath
    });
    return {
      exit_code: 0,
      stdout: [
        `Analysis Input: ${result.analysis_input_path}`,
        `Prompt: ${result.prompt_path}`,
        `Report: ${result.report_path}`,
        `Run Metrics: ${result.run_metrics_path}`,
        ""
      ].join("\n"),
      stderr: ""
    };
  } catch (error) {
    throw pipelineError("build-analysis-input", error);
  }
}

export async function applyMemoryPipeline(options: {
  configPath: string;
  summaryPath: string;
}): Promise<PipelineResult> {
  try {
    const result = await applyMemoryUpdate({
      configPath: options.configPath,
      summaryPath: options.summaryPath
    });
    return {
      exit_code: 0,
      stdout: [
        `Memory updated: ${result.memory_updates}`,
        `Summary: ${result.summary_path}`,
        `MEMORY_UPDATE: ${result.memory_update_path}`,
        `Run Metrics: ${result.run_metrics_path}`,
        `Memory Audit: ${result.memory_audit_path}`,
        ""
      ].join("\n"),
      stderr: ""
    };
  } catch (error) {
    throw pipelineError("apply-memory", error);
  }
}

function pipelineError(action: string, error: unknown): PipelineError {
  const message = error instanceof Error ? error.message : String(error);
  return new PipelineError(`${action} failed: ${message}`, {
    exit_code: 1,
    stdout: "",
    stderr: `${error instanceof Error ? error.name : "Error"}: ${message}\n`
  });
}
